hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/defrag/SnapshotDefragService.java

```diff
@@ -135,48 +196,322 @@ private boolean isRocksToolsNativeLibAvailable() {
+
+          String tableBucketPrefix = bucketPrefixInfo.getTablePrefix(table);
+          String sstFileReaderLowerBound = bucketPrefixInfo.getTablePrefix(entry.getKey());
+          String sstFileReaderUpperBound = null;
+          if (Strings.isNotEmpty(sstFileReaderLowerBound)) {
+            sstFileReaderUpperBound = getLexicographicallyHigherString(sstFileReaderLowerBound);
+          }
+          List<Path> deltaFilePaths = deltaFiles.stream().map(Pair::getKey).collect(Collectors.toList());
+          SstFileSetReader sstFileSetReader = new SstFileSetReader(deltaFilePaths);
+          fileToBeIngested = differTmpDir.resolve(table + "-" + UUID.randomUUID() + SST_FILE_EXTENSION);
+          // Delete all delta files after reingesting into the checkpointStore.
+          filesToBeDeleted.add(fileToBeIngested);
+          int deltaEntriesCount = 0;
+          try (ClosableIterator<String> keysToCheck =
+                   sstFileSetReader.getKeyStreamWithTombstone(sstFileReaderLowerBound, sstFileReaderUpperBound);
+               TableMergeIterator<String, byte[]> tableMergeIterator = new TableMergeIterator<>(keysToCheck,
+                  tableBucketPrefix, snapshotTable, previousSnapshotTable);
+               RDBSstFileWriter rdbSstFileWriter = new RDBSstFileWriter(fileToBeIngested.toFile())) {
+            while (tableMergeIterator.hasNext()) {
+              Table.KeyValue<String, List<byte[]>> kvs = tableMergeIterator.next();
+              // Check if the values are equal or if they are not equal then the value should be written to the
+              // delta sstFile.
+              if (!Arrays.equals(kvs.getValue().get(0), kvs.getValue().get(1))) {
+                try (CodecBuffer key = StringCodec.get().toHeapCodecBuffer(kvs.getKey())) {
+                  byte[] keyArray = key.asReadOnlyByteBuffer().array();
+                  byte[] val = kvs.getValue().get(0);
+                  // If the value is null then add a tombstone to the delta sstFile.
+                  if (val == null) {
+                    rdbSstFileWriter.delete(keyArray);
+                  } else {
+                    rdbSstFileWriter.put(keyArray, val);
+                  }
+                }
+                deltaEntriesCount++;
+              }
+            }
+          } catch (RocksDBException e) {
+            throw new RocksDatabaseException("Error while reading sst files.", e);
+          }
+          if (deltaEntriesCount == 0) {
+            // If there are no delta entries then delete the delta file. No need to ingest the file as a diff.
+            fileToBeIngested = null;
+          }
+        }
+        if (fileToBeIngested != null) {
+          if (!fileToBeIngested.toFile().exists()) {
+            throw new IOException("Delta file does not exist: " + fileToBeIngested);
+          }
+          Table checkpointTable = checkpointStore.getTable(table);
+          checkpointTable.loadFromFile(fileToBeIngested.toFile());
```

Comment:
not sure the use of this. Does loadFromFile() have a side effect? Looks like it's no use. Or do we just want to make sure it does not corrupt and can load?

Reply:
LoadFile if it fails would throw a ROcksDataBaseException the entire defrag would fail